tests/unit/utils/test_sqlalchemy_exec_utils.py

```diff
@@ -62,13 +62,45 @@ def execute(self, query):
 ERROR = OperationalError("A bad thing happend", {}, None)
 
 
+def test_execute_with_retry():
+    tracker = {"error": True, "finished": []}
+
+    def log(msg):
+        tracker["logs"] = tracker.get("logs", []) + [msg]
+
+    num_parts = 4
+
+    def execute():  # Return an iterator to mimic Connection.execute
+        # Succeed for 0, 1, fail for 2 on the first go, then succeed for 2, 3
+        def execute_part(part, tr):
+            if "error" in tr and part == 2:
+                tr.pop("error")
+                raise ERROR
+            return part
+
+        return (execute_part(part, tracker) for part in range(num_parts))
+
+    execute_with_retry = execute_with_retry_factory(
+        execute, max_retries=3, retry_sleep=0, backoff_factor=2, log=log
+    )
+
+    # Do something with the results
+    for part in execute_with_retry():
+        tracker["finished"].append(part)
+
+    assert tracker["finished"] == list(range(num_parts))
+    assert "Retrying query (attempt 1 / 3)" in tracker["logs"]
+    assert "Retrying query (attempt 2 / 3)" not in tracker["logs"]
+
+
 @mock.patch("time.sleep")
-def test_execute_with_retry(sleep):
+def test_sleep_between_retries(sleep):
     execute = mock.Mock(side_effect=[ERROR, ERROR, ERROR, "its OK now"])
```

Comment:
I think there's a slight simpler way of testing the behaviour here which is to replace one of the "immediate error" cases with an iterator that triggers an error part way through e.g.
```suggestion
    def iterate_with_error():
        yield 1
        yield 2
        raise ERROR

    execute = mock.Mock(side_effect=[ERROR, iterate_with_error(), ERROR, ["its", "OK", "now"])
```

So maybe modifying this test to cover that case and adding your log-message-asserting code would be a simpler way to go?